import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/backupuploader/GcpBackupUploaderTest.java

```diff
@@ -0,0 +1,209 @@
+import com.google.cloud.storage.testing.RemoteStorageHelper;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import java.io.ByteArrayInputStream;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.truth.Truth.assertThat;
+
+public class GcpBackupUploaderTest {
+    private static GcpBackupUploader gcpBackupUploader;
+    private static Storage localStorage;
+    private static byte[] BYTE_SEQ_1 = new byte[]{0, 1, 2, 5, 4, 3, 5, 4, 2, 0, 4, 5, 4, 7};
+    private static byte[] BYTE_SEQ_2 = new byte[]{1, 5, 7, 3, 5, 4, 0, 1, 3};
+    private static byte[] BYTES_SEQ = new byte[]{0, 1, 2, 5, 4, 3, 5, 4, 2, 0, 4, 5, 4, 7};
+
+    private static final String UNIQUE_FILE_NAME_1 = "uniq1";
+    private static final String UNIQUE_FILE_NAME_2 = "uniq2";
+    private static final String NOT_EXISTS_UPLOAD_PATH = "gs://testing-healthcare-adapter/some-backup";
+    private static final String UNIQ_NAME = "uniq";
+    private static final String UNIQ_NAME_REMOVE = "uniq_remove";
+    private static final String ENV_CREDS = "GOOGLE_APPLICATION_CREDENTIALS";
+    private static final String AUTH_TYPE = "OAuth2";
+    private static final String UPLOAD_PATH = "gs://testing-healthcare-adapter/test-backup";
+    private static final String GCP_PROJECT_ID = "dev-idg-uvs";
+    private static final String OAUTHSCOPES = "https://www.googleapis.com/auth/cloud-healthcare";
+    private static final String UPLOAD_PATH_EMPTY_BUCKET_NAME = "gs:/// ";
+    private static final String UPLOAD_PATH_SPACE_BUCKET_NAME = "gs:// / ";
+    private static final String UPLOAD_PATH_EMPTY_UPLOAD_OBJECT = "gs:///some//";
+    private static final String UPLOAD_PATH_SPACE_UPLOAD_OBJECT = "gs://some/ ";
+    private static final String PROJECT_NAME = "dev-idg-uvs";
+    private static final String BUCKET_NAME = "testing-healthcare-adapter";
+    private static final String UPLOAD_OBJECT = "test-backup";
+
+    /**
+    * To create an instance of the Storage class, the tests used RemoteStorageHelper
+    * because LocalStorageHelper returns an "method not implemented yet" error when
+    * trying to write data to the bucket. The tests use a temporary bucket that is
+    * deleted after all tests have run
+     */
+    @BeforeClass
+    public static void setUp() {
+        try {
+            RemoteStorageHelper helper =
+                    RemoteStorageHelper.create(PROJECT_NAME, new FileInputStream(System.getenv(ENV_CREDS)));
+            localStorage = helper.getOptions().getService();
+            localStorage.create(BucketInfo.of(BUCKET_NAME));
+            gcpBackupUploader = new GcpBackupUploader(UPLOAD_PATH, localStorage);
+        } catch (IOException e) {
+            e.printStackTrace();
```

Comment:
Since this project is open source, anyone should be able to run the unit tests. I think dialing into a remote storage on a cloud project won't really work for us, e.g. imagine two users run these tests in parallel, they may fail. Also this requires us to give all users access to a remote project and bucket. Can we re-work these tests so that they don't require using remote GCS storage? 

One solution could be to make the backup uploader take in an interface for interacting with GCS, and then mock out that interface ourselves.

Reply:
This approach to create Storage was used on the recommendation of Google developers -> https://github.com/googleapis/java-storage#quickstart. It is impossible to use the LocalStorageHelper class for our tests, since it does not implement the function of emulating data writing. Immediately after the end of the tests, the temporary Bucket is deleted and will be created again for the next user. This gives us a guarantee of the correct operation and the deletion of temporary data

Reply:
As a solution, I propose to use the option of creating a Storage object using the default settings, the user will not need to create ENVs and do not need to specify credentials, also Bucket id will be generated randomly based on UUIDs, where the user can run tests at the same time

Reply:
Hmm, but I also don't think we want this project to depend on a bucket in an external project, and give users permissions to write to that project. They could write whatever they want to those buckets with the credentials right?

Reply:
All tests are executed in a temporary Bucket, which is removed immediately. No existing Bucket will be affected or changed. Also, the user does not know the name of the Bucket and does not affect the process of its creation, modification or allocation.There is one bottleneck - the user must have write permissions. But if he uses the option with gcp backup, then he must have such permissions for the backup to work.

Reply:
Yeah, but still the user can modify this code to write whatever they want to your bucket right? Also linking this project with an external bucket and leaving the credentials in the text file is a bit worrisome. And if we don't leave the credentials in the file, no one will be able to run this test.

I'm fine with deleting this test altogether rather than using an external bucket and remote storage, but I think creating an interface for interacting with GCS, similar to how we create a DicomWebClient and mock it in other tests, would be the ideal approach.

Reply:
local FakeStorage implemented by Nikolai